example/src/main/scala/AuthenticationAdvanced.scala

```diff
@@ -0,0 +1,78 @@
+import zhttp.http.auth._
+import zhttp.http.{ Method, _ }
+import zhttp.service.Server
+import zio._
+
+import java.util.Base64
+import scala.util.Try
+
+object AuthenticationAdvanced extends App {
+
+  case class AuthInfo(username: String)
+
+  type AuthMiddleware = HttpAuthenticationMiddleware[Any, AuthInfo]
+
+  private val users: Map[String, String] = Map(
+    "John De Goes" -> "ZIO rocks!",
+    "Adam Fraser" -> "I love ZIO!"
+  )
+
+  def authenticate(request: Request): Either[HttpError, AuthInfo] =
+    for {
+      headerValue <- request.getAuthorization.toRight(HttpError.Unauthorized())
+      value <- headerValue.split(" ").map(_.trim).toList match {
+        case scheme :: value :: Nil if scheme.equalsIgnoreCase("basic") => Right(value)
+        case _ => Left(HttpError.Unauthorized())
+      }
+      decoded <- Try { new String(Base64.getDecoder.decode(value)) }
+                    .toOption
+                    .toRight(HttpError.Unauthorized())
+      userNamePassword <- decoded.split(":").map(_.trim).toList match {
+        case usr :: pass :: Nil => Right((usr -> pass))
+        case _ => Left(HttpError.Unauthorized())
+      }
+      (user, password) = userNamePassword
+      authenticated <- users.collectFirst {
+        case (usr, pass) if usr.equalsIgnoreCase(user)
+          && pass.equalsIgnoreCase(password) => usr
+      } toRight(HttpError.Forbidden())
+    } yield AuthInfo(authenticated)
```

Comment:
Do we need such an example for auth in user-land code? It's a little intimidating for first-timers 🙈

Reply:
yeah we can simplify, but it's just a simple basic auth :). I'll simplify.